Use variable for ocp version detection

Vincent056 opened the pull request:
In order to use OCP version detection on HyperShift we need to introduce variables to specify the API path.

We have introduced the rule [version_detect/rule.yml](https://github.com/ComplianceAsCode/content/pull/9726/files#diff-dafef22e37ea492d6725a6c29ab6db99dd0941322150768063f6fd3d9c241164), this rule is used to fetch ocp version api resource, user can set [ocp_version_yaml_path.var](https://github.com/ComplianceAsCode/content/pull/9726/files#diff-f739a3aaf4cb3269532a94bfb8c73532be05dbaa9fa2c376c3a87c8f0c9ebfd8) and [applications/openshift/ocp_version_api_path.var](https://github.com/ComplianceAsCode/content/pull/9726/files#diff-612fb661ba294f5a0823e7b09d814ce672d6582c3d666ffab7c2f727bc57b6fb) to set custom api path for a HyperShift cluster.

Example on how you can you can set HyperShift cluster name in a tailoredProfile so that the version detection will detect the OCP version of HyperShift cluster correctly.

```yaml
apiVersion: compliance.openshift.io/v1alpha1
kind: TailoredProfile
metadata:
  name: hypershift-cis
spec:
  description: scan a hypershift cluster
  setValues:
  - name: upstream-ocp4-hypershift-cluster
    rationale: Scan this cluster
    value: example-guest-cluster
  extends: upstream-ocp4-cis
  title: CIS
```

ggbecker commented on shared/checks/oval/installed_app_is_ocp4.xml:
```diff
@@ -191,7 +192,7 @@
 
   <local_variable id="hypershift_hosted_cluster_location" datatype="string" comment="The actual filepath of the file to scan." version="1">
-    <literal_component>/kubernetes-api-resources/apis/apiextensions.k8s.io/v1/customresourcedefinitions/hostedclusters.hypershift.openshift.io</literal_component>
+    <literal_component>/kubernetes-api-resourceso</literal_component>
```
```suggestion
    <literal_component>/kubernetes-api-resources</literal_component>
```

jhrozek commented on applications/openshift/helper/version_detect/rule.yml:
```diff
@@ -0,0 +1,22 @@
+
+warnings:
+- general: |-
+    {{{ openshift_filtered_version({custom_api_path: dump_path}) | indent(4) }}}
```
Making sure that I understand this part and the jinja part correctly: Does this say that we will save `custom_api_path` filtered through `custom_jqfilter` but regardless of their actual value, we will pretend that they come from `/apis/config.openshift.io/v1/clusteroperators/openshift-apiserver` ?

jhrozek replied:
By pretend, I mean save in the data dump at location normally used for `/apis/config.openshift.io/v1/clusteroperators/openshift-apiserver` ?

Vincent056 replied:
that's correct! And we just need to adjust those two variables in a HyperShift Env

Vincent056 replied:
that's correct! And we just need to adjust those two variables in a HyperShift Env

Vincent056 replied:
just changed the data dump location to `/ocp/version` to be backward consistent

jhrozek commented on applications/openshift/ocp_version_yaml_path.var:
```diff
@@ -0,0 +1,21 @@
+interactive: false
+
+options:
+    default: "[.status.versions[].version]"
```
Why do we have the square braces around the path? Is it because `[shared/checks/oval/installed_app_is_ocp4.xml](https://github.com/ComplianceAsCode/content/pull/9726/commits/563a00367342be6f7d5ed17d2c27add3d0a1cc89#diff-79f1a46adfc289cd2d6b69c96af6c3a98e38ba749db957fcb7c823ac1a4430f6)` uses a list expansion so we need to force a list here?

Vincent056 replied:
Actually, I am not sure if we need the square braces, let me try without it

jhrozek reviewed (commented):
It seems that the `nerc-cip` profile also needs the rule added.
So far I've just added comments to make sure I understand the code. I'm going to install hypershift again and see how its output looks like.
btw the e2e failures look legit, but I don't know why those rules fail...

Vincent056 commented:
> It seems that the `nerc-cip` profile also needs the rule added. So far I've just added comments to make sure I understand the code. I'm going to install hypershift again and see how its output looks like. btw the e2e failures look legit, but I don't know why those rules fail...

We need to make change to CO in order to for this to pass, because we changed the jq filter and yaml path

Vincent056 commented:
> It seems that the `nerc-cip` profile also needs the rule added. So far I've just added comments to make sure I understand the code. I'm going to install hypershift again and see how its output looks like. btw the e2e failures look legit, but I don't know why those rules fail...

We need to make a change to CO api resource collector  in order to for this to pass, because we changed the jq filter and yaml path

Vincent056 commented:
This works in conjunction with https://github.com/ComplianceAsCode/compliance-operator/pull/176

Vincent056 commented:
> It seems that the `nerc-cip` profile also needs the rule added. So far I've just added comments to make sure I understand the code. I'm going to install hypershift again and see how its output looks like. btw the e2e failures look legit, but I don't know why those rules fail...

nerc-cip will be extending moderate profile

Vincent056 commented:
/retest

Vincent056 commented:
/retest

Vincent056 commented:
/retest

jhrozek commented:
/test e2e-aws-rhcos4-high
the cluster for the high profile did not install

jhrozek reviewed (approved):
LGTM
thank you, this works for me. I used the following `TailoredProfile` to test:
```
apiVersion: compliance.openshift.io/v1alpha1
kind: TailoredProfile
metadata:
  name: hypershift-version
spec:
  description: Scans a hypershift cluster
  setValues:
  - name: upstream-ocp4-hypershift-cluster
    rationale: Scan this cluster
    value: jhrozek-guest-cluster
  extends: upstream-ocp4-cis
  title: CIS for hypershift guest cluster
```

My environment was a 4.11 host cluster and a 4.12 guest cluster. Then I modified one of the rules locally to be applicable to 4.12 only and ran a scan with the above TP. The rule was passing, when I modified the rule to be 4.11 only, it was NOT-APPLICABLE as expected.

Thank you for the patience during the review and the nice work!

jhrozek commented:
The unit test failures are irrelevant:
```
---------Bash Remediation Logs---------
ERROR - No tests found matching the rule ID(s) 'version_detect_in_hypershift, version_detect_in_ocp'
WARNING - Nothing has been tested!
---------Ansible Remediation Logs---------
ERROR - No tests found matching the rule ID(s) 'version_detect_in_hypershift, version_detect_in_ocp'
WARNING - Nothing has been tested!
```

Vincent056 commented:
> LGTM thank you, this works for me. I used the following `TailoredProfile` to test:
> 
> ```
> apiVersion: compliance.openshift.io/v1alpha1
> kind: TailoredProfile
> metadata:
>   name: hypershift-version
> spec:
>   description: Scans a hypershift cluster
>   setValues:
>   - name: upstream-ocp4-hypershift-cluster
>     rationale: Scan this cluster
>     value: jhrozek-guest-cluster
>   extends: upstream-ocp4-cis
>   title: CIS for hypershift guest cluster
> ```
> 
> My environment was a 4.11 host cluster and a 4.12 guest cluster. Then I modified one of the rules locally to be applicable to 4.12 only and ran a scan with the above TP. The rule was passing, when I modified the rule to be 4.11 only, it was NOT-APPLICABLE as expected.
> 
> Thank you for the patience during the review and the nice work!

Thanks for your reivews, and testing it out the patch!

jhrozek reviewed (approved):
retested with the latest version of https://github.com/ComplianceAsCode/compliance-operator/pull/176 and everything works fine

ggbecker commented:
/packit retest-failed

tmishina commented:
@jhrozek Thank you very much for supporting HyperShift environment!
Do you know when will this PR available for the Compliance Operator installed by OperatorHub?

The ComplianceAsCode/content image that current version (v0.1.57) of Compliance Operator is `registry.redhat.io/compliance/openshift-compliance-content-rhel8@sha256:d78ad0a8ebb5fc2954a0e6d5cfa615a48ba4875eafc64742d1509d8ad509831e`, and it does not contain a variable `ocp4-hypershift-cluster`.

The tailored profile resource I used for my test is as follows.

```yaml
apiVersion: compliance.openshift.io/v1alpha1
kind: TailoredProfile
metadata:
  name: hypershift-version
  namespace: openshift-compliance
spec:
  description: Scans a hypershift cluster
  extends: ocp4-cis
  setValues:
  - name: ocp4-hypershift-cluster
    rationale: Scan this cluster
    value: jhrozek-guest-cluster
  title: CIS for hypershift guest cluster
status:
  errorMessage: 'fetching variable: Variable.compliance.openshift.io "ocp4-hypershift-cluster"
    not found'
  outputRef:
    name: ""
    namespace: ""
  state: ERROR
```

Vincent056 commented:
@tmishina looks like `https://github.com/ComplianceAsCode/compliance-operator/pull/176` patch was merged after v0.1.59, so I think it will be available on v0.1.60.
